controllers/config/clientconfig_controller.go

```diff
@@ -0,0 +1,215 @@
+		r.shutdownFunc()
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager will only set this controller to listen in control plane cluster
+func (r *ClientConfigReconciler) SetupWithManager(mgr ctrl.Manager, cancelFunc context.CancelFunc) error {
+	r.shutdownFunc = cancelFunc
+	if r.secretFilter == nil {
+		r.secretFilter = make(map[types.NamespacedName]types.NamespacedName)
+	}
+
+	// We should only reconcile objects that match the rules
+	toMatchingClientConfig := func(secret client.Object) []reconcile.Request {
+		requests := []reconcile.Request{}
+		secretKey := types.NamespacedName{Name: secret.GetName(), Namespace: secret.GetNamespace()}
+		if clientConfigName, found := r.secretFilter[secretKey]; found {
+			requests = append(requests, reconcile.Request{NamespacedName: clientConfigName})
+		}
+		return requests
+	}
+
+	cb := ctrl.NewControllerManagedBy(mgr).
+		For(&configapi.ClientConfig{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(&source.Kind{Type: &corev1.Secret{}}, handler.EnqueueRequestsFromMapFunc(toMatchingClientConfig))
+
+	return cb.Complete(r)
+}
+
+// InitClientConfigs will fetch clientConfigs from the current cluster (control plane cluster) and create all the required Cluster objects for
+// other controllers to use. Not called from SetupWithManager since other controllers need the []cluster.Cluster array
+func (r *ClientConfigReconciler) InitClientConfigs(ctx context.Context, mgr ctrl.Manager, watchNamespace string) ([]cluster.Cluster, error) {
+	logger := log.FromContext(ctx)
+
+	uncachedClient := r.ClientCache.GetLocalNonCacheClient()
+	clientConfigs := make([]configapi.ClientConfig, 0)
+	namespaces := strings.Split(watchNamespace, ",")
+
+	for _, ns := range namespaces {
+		cConfigs := configapi.ClientConfigList{}
+		err := uncachedClient.List(ctx, &cConfigs, client.InNamespace(ns))
+		if err != nil {
+			return nil, err
+		}
+		clientConfigs = append(clientConfigs, cConfigs.Items...)
+	}
+
+	additionalClusters := make([]cluster.Cluster, 0, len(clientConfigs))
+
+	// TODO Secret could point to multiple clientConfigs. Shouldn't matter in our current use-case
+	r.secretFilter = make(map[types.NamespacedName]types.NamespacedName, len(clientConfigs))
+
+	for _, cCfg := range clientConfigs {
+		// Calculate hashes
+		cCfgName := types.NamespacedName{Name: cCfg.Name, Namespace: cCfg.Namespace}
+		secretName := types.NamespacedName{Name: cCfg.Spec.KubeConfigSecret.Name, Namespace: cCfg.Namespace}
+
+		cCfgHash, secretHash, err := calculateHashes(ctx, uncachedClient, cCfg)
+		if err != nil {
+			return nil, err
+		}
+
+		metav1.SetMetaDataAnnotation(&cCfg.ObjectMeta, ClientConfigHashAnnotation, cCfgHash)
+		metav1.SetMetaDataAnnotation(&cCfg.ObjectMeta, KubeSecretHashAnnotation, secretHash)
+
+		if err := uncachedClient.Update(ctx, &cCfg); err != nil {
+			return nil, err
+		}
+
+		// Add the Secret to the cache
+		r.secretFilter[secretName] = cCfgName
```

Comment:
Why do we need this cache? I understand that different ClientConfigs can point to the same secret. The secret needs to be in the same namespace as the ClientConfig and we're not managing the secrets. In particular if a ClientConfig is deleted, we aren't deleting the secret it references.

Reply:
Because we can't necessarily manage the secrets, so we need a way to manage which secrets we monitor. If that secret is modified, we have to link that to the ClientConfig which we would reconcile.